src/backend/marsha/core/api.py

```diff
@@ -678,6 +687,66 @@ def initiate_upload(self, request, pk=None):
         return Response(response)
 
 
+class LiveRegistrationViewSet(
+    ObjectPkMixin,
+    mixins.CreateModelMixin,
+    mixins.RetrieveModelMixin,
+    viewsets.GenericViewSet,
+):
+    """Viewset for the API of the LiveRegistration object."""
+
+    permission_classes = [permissions.IsVideoToken]
+    queryset = LiveRegistration.objects.all()
+    serializer_class = serializers.LiveRegistrationSerializer
+
+    def get_queryset(self):
+        """Restrict list access to liveRegistration related to the video in the JWT token."""
+        user = self.request.user
+        if isinstance(user, TokenUser) and user.token.payload is not None:
+            # admin and instructors can access all registrations
+            if user.token.payload.get("roles") and any(
+                role in ["administrator", "instructor"]
+                for role in user.token.payload["roles"]
+            ):
+                return LiveRegistration.objects.filter(
+                    id=self.kwargs["pk"], video__id=user.id
+                )
+            # others can only read their registrations
+            if user.token.payload.get("user"):
+                if user.token.payload["user"].get("email"):
+                    # token has an email
+                    return LiveRegistration.objects.filter(
+                        id=self.kwargs["pk"],
+                        video__id=user.id,
+                        email=user.token.payload["user"]["email"],
+                    )
+                # no email, user has access to this registration if it's the right combinaison of
+                # lti_user_id and lti_consumer_id
+                if user.token.payload["user"].get("id") and user.token.payload.get(
+                    "context_id"
+                ):
+                    playlist = Playlist.objects.get(
+                        lti_id=user.token.payload["context_id"]
+                    )
+                    return LiveRegistration.objects.filter(
+                        id=self.kwargs["pk"],
+                        lti_consumer_site=playlist.consumer_site,
+                        lti_user_id=user.token.payload["user"]["id"],
+                        video__id=user.id,
+                    )
+
+        return LiveRegistration.objects.none()
+
+    def create(self, request, *args, **kwargs):
+        """Handle exception on create."""
+        try:
+            return super().create(request, *args, **kwargs)
+        except IntegrityError:
+            return Response(
+                {"error": "IntegrityError"}, status=status.HTTP_400_BAD_REQUEST
+            )
```

Comment:
I don't think this is a good idea. In the validation phase, you should make sure that the query is correct and should succeed. If after this validation, you still get an integrity error, then the 500 error should be raised :man_shrugging:

Reply:
Do you consider that returning a `400` with the `integrity` error disclose a sensitive information ? Otherwise we should deal with this error in the front application and the 500 error should never be raised.

Reply:
In the front application we will have a form with an input asking to choose an email. If this email is already registered what should happen ? IMO we must inform the user that the email is already used and we should ask for an other one.

Reply:
IntegrityError is too large and could be caused by many things. So for the situations that we can already predict will happen, we need to anticipate them during validation and return a helpful message in a 400 error. Other cases of IntegrityErrors are unforseen and should be left as 500 errors... independently of how they are notified to the user in the Ux?

Reply:
Yes I agree, not all `IntegrityError` should be caught. So the email should be tested in the serializer ?

Reply:
Yes I would do that.

Reply:
To speak about it, I've moved this try-catch on integrity error to the serializer and still raise a 400 but I didn't parse errors more than I did before: checking if the video is scheduled or email is the one expected. For all the other integrity errors, I would like you to give me your feedback on how to deal with them.
Should I?
- leave the integrity error and get rid of this try-catch in the serializer?
- try to get the already existing record by querying liveRegistration to identify the specific case and give a better message?
- search in the error string catch in the try-catch to find the name of the IntegrityError's key, example liveregistration_unique_video_lti_idx to give a specific message ?

thanks

Reply:
test